unit test TestCertsTLS fails when IPV6 is disabled

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT 

**Kubernetes version** (use `kubectl version`):
community master

**Environment**:
- **Cloud provider or hardware configuration**: my local machine
- **OS** (e.g. from /etc/os-release): linux
- **Kernel** (e.g. `uname -a`): Linux localhost.localdomain 4.10.8-200.fc25.x86_64 #1 SMP Fri Mar 31 13:20:22 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**:
- **Others**:
ipv6 is disabled in my environment.
net.ipv6.conf.all.disable_ipv6=0

**What happened**:
`$ make test WHAT=k8s.io/kubernetes/federation/pkg/kubefed/init KUBE_GOFLAGS="-v" KUBE_TEST_ARGS='-run ^TestCertsTLS$'
Running tests for APIVersion: v1,admissionregistration.k8s.io/v1alpha1,admission.k8s.io/v1alpha1,apps/v1beta1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2alpha1,batch/v1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,imagepolicy.k8s.io/v1alpha1,networking.k8s.io/v1,policy/v1beta1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,federation/v1beta1
+++ [0611 16:37:17] Running tests without code coverage
=== RUN   TestCertsTLS
--- FAIL: TestCertsTLS (5.11s)
	init_test.go:522: [0] unexpected error: dial tcp [::]:41981: connect: cannot assign requested address
	init_test.go:522: [1] unexpected error: dial tcp [::]:45949: connect: cannot assign requested address
	init_test.go:522: [2] unexpected error: dial tcp [::]:42481: connect: cannot assign requested address
	init_test.go:522: [3] unexpected error: dial tcp [::]:43051: connect: cannot assign requested address
	init_test.go:522: [4] unexpected error: dial tcp [::]:38445: connect: cannot assign requested address
	init_test.go:522: [5] unexpected error: dial tcp [::]:45791: connect: cannot assign requested address
	init_test.go:511: [6] unexpected error: want x509.HostnameError, got: *net.OpError
	init_test.go:511: [7] unexpected error: want x509.HostnameError, got: *net.OpError
	init_test.go:511: [8] unexpected error: want x509.HostnameError, got: *net.OpError
	init_test.go:511: [9] unexpected error: want x509.HostnameError, got: *net.OpError
	init_test.go:515: [10] unexpected error: want x509.UnknownAuthorityError, got: *net.OpError
	init_test.go:515: [11] unexpected error: want x509.UnknownAuthorityError, got: *net.OpError
	init_test.go:511: [12] unexpected error: want x509.HostnameError, got: *net.OpError
	init_test.go:511: [13] unexpected error: want x509.HostnameError, got: *net.OpError
FAIL
exit status 1
FAIL	k8s.io/kubernetes/federation/pkg/kubefed/init	5.391s
Makefile:164: recipe for target 'test' failed
make: *** [test] Error 1`


**What you expected to happen**:
Unit test should passes.

**How to reproduce it** (as minimally and precisely as possible):
This always happens in my environment.

**Anything else we need to know**:

--- a/federation/pkg/kubefed/init/init_test.go
+++ b/federation/pkg/kubefed/init/init_test.go
@@ -1444,7 +1444,11 @@ func tlsHandshake(t *testing.T, sCfg, cCfg *tls.Config) error {
 		}
 	}()
 
-	c, err := tls.Dial("tcp", s.Addr().String(), cCfg)
+	// workaround [::] not working in ipv4 only systems (https://github.com/golang/go/issues/18806)
+	// TODO: remove with Golang 1.9 with https://go-review.googlesource.com/c/45088/
+	addr := strings.TrimPrefix(s.Addr().String(), "[::]")
+
+	c, err := tls.Dial("tcp", addr, cCfg)
 	if err != nil {
 		// Intentionally not serializing the error received because we want to
 		// test for the failure case in the caller test function.

